File: donghwa/0x08/4949.py
from collections import deque
import sys


while(1) :
    line = sys.stdin.readline().rstrip()
    if line == "." :
        break
    S = deque([])
    check = True
    for c in line:
        if (c=='(' or c=='[') :
            S.append(c)
        elif (c == ')') :
            if (len(S)==0 or S[-1] != '('):
                check = False
                break
            S.pop()
        elif (c == ']') :
            if (len(S)==0 or S[-1] != '['):
                check = False
                break
            S.pop()
    if (len(S)>0) :
        check=False
    if (check) :
        print("yes")
    else :
        print("no")

File: donghwa/0x08/4949.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);


    while(true){
        string line;
        getline(cin, line);
        // cout << "line :" << line << '\n';
        if (line == ".") break;
        stack<char> S;
        bool check = true;
        for (auto c : line){
            if (c == '(' || c == '[') S.push(c);
            else if (c == ')'){
                if (S.empty() || S.top() != '('){
                    check=false; break;
                }
                S.pop();
            }
            else if (c == ']'){
                if (S.empty() || S.top() != '['){
                    check = false; break;
                }
                S.pop();
            }
        }
        if (!S.empty()) check=false;
        if (check) cout << "yes\n";
        else cout << "no\n";
    }
}